src/Omega_h_adios2.cpp

```diff
@@ -0,0 +1,563 @@
+    read_array(io, reader, gclas_id,name);
+
+    for (int32_t j = 0; j < npairs; ++j) {
+      ClassPair pair;
+      pair.dim=gclas_dim[j];
+      pair.id=gclas_id[j];
+      mesh->class_sets[gclas_name].push_back(pair);
+    }
+  }
+}
+
+void write_parents(adios2::IO &io, adios2::Engine &writer, Mesh* mesh, std::string pref)
+{
+  int32_t has_parents = mesh->has_any_parents();
+  std::string name = pref+"has_parents";
+  write_value(io, writer, has_parents, name);
+  if (has_parents) 
+  {
+    for (int32_t d = 0; d <= mesh->dim(); ++d) 
+    {
+      auto parents = mesh->ask_parents(d);
+      name = pref+"parent_"+to_string(d)+"_idx";
+      write_array(io, writer, parents.parent_idx, name);
+      name = pref+"parent_"+to_string(d)+"_codes";
+      write_array(io, writer, parents.codes, name);
+    }
+  }
+}
+
+void read_parents(adios2::IO &io, adios2::Engine &reader, Mesh* mesh, std::string pref)
+{
+  int32_t has_parents;
+  std::string name = pref+"has_parents";
+  read_value(io, reader, &has_parents, name);
+  if (has_parents) 
+  {
+    for (int32_t d = 0; d <= mesh->dim(); ++d) 
+    {
+      Parents parents;
+      name = pref+"parent_"+to_string(d)+"_idx";
+      read_array(io, reader, parents.parent_idx, name);
+      name = pref+"parent_"+to_string(d)+"_codes";
+      read_array(io, reader, parents.codes, name);
+      mesh->set_parents(d, parents);
+    }
+  }
+}
+
+void _write_adios2(adios2::IO &io, adios2::Engine & writer,
+                  Mesh* mesh, std::string pref)
+{
+  comm_size = mesh->comm()->size();
+  rank = mesh->comm()->rank();
+
+  write_meta(io, writer, mesh, pref);
+  int32_t nverts = mesh->nverts();
+  std::string name=pref+"nverts";
+  write_value(io, writer, nverts, name);
+
+  for (int32_t d = 1; d <= mesh->dim(); ++d)
+    write_down(io, writer, mesh, d, pref);
+
+  for (Int d = 0; d <= mesh->dim(); ++d)
+  {
+    write_tags(io, writer, mesh, d, pref);
+    write_pbdry(io, writer, mesh, d, pref);
+  }
+
+  write_sets(io, writer, mesh, pref);
+  write_parents(io, writer, mesh, pref);
+}
+
+void write_adios2(filesystem::path const& path, 
+		  std::map<Mesh*, std::string>& mesh_map)
+{
+  std::map<Mesh*, std::string>::iterator it=mesh_map.begin();
+  Mesh* mesh=it->first;
+
+  if (path.extension().string() != ".bp" && can_print(mesh)) {
+    std::cout
+        << "it is strongly recommended to end Omega_h paths in \".bp\",\n";
+    std::cout << "instead of just \"" << path << "\"\n";
+  }
+  filesystem::create_directory(path);
+
+  adios2::ADIOS adios(mesh->comm()->get_impl());
+  adios2::IO io = adios.DeclareIO("mesh-writer");
+  std::string filename=path.c_str();
+
+  adios2::Engine writer = io.Open(filename, adios2::Mode::Write);
+  writer.BeginStep();
+  for (; it!=mesh_map.end(); ++it)
+    _write_adios2(io, writer, it->first, it->second);
+  writer.EndStep();
+  writer.Close();
+}
+
+
+void write_adios2(filesystem::path const& path, Mesh* mesh, std::string pref)
+{
```

Comment:
I think you have your api backwards which is causing some issues. I don't think `_write_adios2` should exist, but instead, this function should write to the specified path. If you want to have the function signature that takes a map it will just loop through and call this function multiple times like you have done with `write_adios2`.